File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.java
class Solution {

    private static final int EMPTY = 0;
    private static final int PATH = 1;
    private static final int[][] OFFSET = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    private int N, M;

    public int solution(int[][] board, int[] aloc, int[] bloc) {
        N = board.length;
        M = board[0].length;

        return dfs(board, aloc[0], aloc[1], bloc[0], bloc[1]).count;
    }

    private Result dfs(int[][] board, int aY, int aX, int bY, int bX) {
        if (board[aY][aX] == EMPTY) {
            return new Result(false, 0);
        }

        boolean win = false;
        int minTurn = Integer.MAX_VALUE;
        int maxTurn = 0;

        for (int dir = 0; dir < 4; dir++) {
            int aY_ = aY + OFFSET[dir][0];
            int aX_ = aX + OFFSET[dir][1];
            if (isIn(aY_, aX_) && board[aY_][aX_] == PATH) {
                board[aY][aX] = EMPTY;
                Result next = dfs(board, bY, bX, aY_, aX_);
                board[aY][aX] = PATH; 

                if (!next.win) {
                    win = true;
                    minTurn = Math.min(minTurn, next.count + 1);
                } else if (!win) {
                    maxTurn = Math.max(maxTurn, next.count + 1);
                }
            }
        }

        if (win) return new Result(true, minTurn);
        else     return new Result(false, maxTurn);
    }

    private boolean isIn(int y, int x) {
        return 0 <= y && y < N && 0 <= x && x < M;
    }

    static class Result {
        boolean win;
        int count;

        Result(boolean win, int count) {
            this.win = win;
            this.count = count;
        }
    }
    
}

File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.cpp
#include <vector>
#include <algorithm>

using namespace std;

const int EMPTY = 0;
const int PATH  = 1;
const int OFFSET[4][2] = {{1,0},{0,1},{-1,0},{0,-1}};

int N, M;

bool isIn(int y, int x) { return 0 <= y && y < N && 0 <= x && x < M; }

// {이길 수 있는지, 이동 횟수}
pair<bool, int> dfs(vector<vector<int>>& board, int aY, int aX, int bY, int bX) {
    // 내 차례인데 발판이 없으면 패배
    if (board[aY][aX] == EMPTY) {
        return {false, 0};
    }
    
    bool win = false;
    int minTurn = 1e9, maxTurn = 0;

    for (int dir = 0; dir < 4; dir++) {
        int aY_ = aY + OFFSET[dir][0];
        int aX_ = aX + OFFSET[dir][1];
        if (isIn(aY_, aX_) && board[aY_][aX_] == PATH) {
            board[aY][aX] = EMPTY;
            auto [oppWin, turns] = dfs(board, bY, bX, aY_, aX_);
            board[aY][aX] = PATH;

            // 상대가 못이기면, 내가 이기는 경우
            if (!oppWin) {      
                win = true;
                minTurn = min(minTurn, turns + 1);
            } 
            // 이미 이길 수 있는 방법이 있으면 아래는 안 봐도 됨
            else if (!win) {  
                maxTurn = max(maxTurn, turns + 1);
            }
        }
    }
    
    // 이길 수 있으면 최소 턴, 아니면 최대 턴
    if (win) return {true, minTurn};
    return {false, maxTurn};
}

int solution(vector<vector<int>> board, vector<int> aloc, vector<int> bloc) {
    N = board.size();
    M = board[0].size();
    
    return dfs(board, aloc[0], aloc[1], bloc[0], bloc[1]).second;
}